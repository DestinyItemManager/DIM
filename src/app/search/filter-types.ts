import { ItemHashTag } from '@destinyitemmanager/dim-api-types';
import { t } from 'app/i18next-t';
import { ItemInfos } from 'app/inventory/dim-item-info';
import { DimItem } from 'app/inventory/item-types';
import { DimStore } from 'app/inventory/store-types';
import { Loadout } from 'app/loadout/loadout-types';
import { InventoryWishListRoll } from 'app/wishlists/wishlists';
type I18nInput = Parameters<typeof t>;

// a filter can return various bool-ish values
type ValidFilterOutput = boolean | null | undefined;

export type ItemFilter = (item: DimItem) => ValidFilterOutput;

/**
 * A slice of data that could be used by filter functions to
 * initialize some data required by particular filters. If a new filter needs
 * context that isn't here, add it to this interface and makeSearchFilterFactory
 * in search-filter.ts.
 */
export interface FilterContext {
  stores: DimStore[];
  allItems: DimItem[];
  currentStore: DimStore;
  loadouts: Loadout[];
  inventoryWishListRolls: { [key: string]: InventoryWishListRoll };
  newItems: Set<string>;
  itemInfos: ItemInfos;
  itemHashTags: {
    [itemHash: string]: ItemHashTag;
  };
  language: string;
}

// TODO: FilterCategory

/**
 * A definition of a filter or closely related group of filters. This is
 * self-contained and can be used for both autocomplete and for building up the
 * filter expression itself. We can also use it to drive filter help and filter
 * editor.
 */
export type FilterDefinition = {
  /**
   * One or more keywords which trigger the filter when typed into search bar.
   * What this means depends on what "format" this filter is.
   */
  keywords: string | string[];

  /**
   * A t()-compatible arg tuple or i18n key pointing to a full description of
   * the filter, to show in filter help
   */
  description: string | I18nInput;

  /**
   * What kind of query this is, used to help generate suggestions.
   *
   * `undefined` - a simple 'is/not' filter. the filter itself knows everything it's looking for
   *
   * `query` - a starting stem and a pre-known value, like "tag:keep". a filterValue will be involved and will match a string we expect
   *
   * `freeform` - a starting stem and a freeform value. the filterValue will be some arbitrary string we test against other strings
   *
   * `range` - a starting stem and a mathlike string afterward like <=5
   *
   * `rangeoverload` - a starting stem like "masterwork" and then either a mathlike string or a word
   *
   * `custom` - suppresses automated suggestion generation so suggestionsGenerator is the only source of suggestions
   */
  format?: 'query' | 'freeform' | 'range' | 'rangeoverload' | 'custom';

  /** destinyVersion - 1 or 2, or if a filter applies to both, undefined */
  destinyVersion?: 1 | 2;

  /** methods for retiring a filter */
  deprecated?: FilterDeprecation;

  /**
   * A function that is given context about the query and the world around it
   * (FilterContext) and should generate a simple filter function that is given
   * an item and returns whether that item should be included in the search.
   * Because this is a function that returns the item filter function, and it is
   * invoked once at the point where we parse the query, it can be used to
   * pre-process information that is needed by the actual filter function. that,
   * given a value from a more complex filter expression and the context about
   * the world around it, can generate a filter function. In that case, the
   * filter function will be generated once, at the point where the overall
   * query is parsed.
   */
  filter: (args: { filterValue: string } & FilterContext) => ItemFilter;

  /**
   * A list of suggested keywords for filters that can take a freeform filter value.
   */
  suggestions?: string[];

  /**
   * A custom function used to generate (additional) suggestions
   */
  // TODO: give this access to a SuggestionsContext arg
  // TODO: add manifest to SuggestionsContext and we can generate archetype/perk/etc suggestions
  suggestionsGenerator?: () => string[];
};

export const enum FilterDeprecation {
  NotDeprecated,
  Deprecated,
  Disabled,
}
