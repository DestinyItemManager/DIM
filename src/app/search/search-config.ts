import { DestinyVersion } from '@destinyitemmanager/dim-api-types';
import { destinyVersionSelector } from 'app/accounts/selectors';
import { createSelector } from 'reselect';
import { canonicalFilterFormats, FilterDefinition, SuggestionsContext } from './filter-types';
import advancedFilters from './search-filters/advanced';
import d1Filters from './search-filters/d1-filters';
import dupeFilters from './search-filters/dupes';
import freeformFilters from './search-filters/freeform';
import itemInfosFilters from './search-filters/item-infos';
import knownValuesFilters from './search-filters/known-values';
import loadoutFilters from './search-filters/loadouts';
import simpleRangeFilters from './search-filters/range-numeric';
import overloadedRangeFilters from './search-filters/range-overload';
import simpleFilters from './search-filters/simple';
import socketFilters from './search-filters/sockets';
import statFilters from './search-filters/stats';
import locationFilters from './search-filters/stores';
import wishlistFilters from './search-filters/wishlist';
import { generateSuggestionsForFilter, suggestionsContextSelector } from './suggestions-generation';

const allFilters = [
  ...dupeFilters,
  ...($featureFlags.wishLists ? wishlistFilters : []),
  ...freeformFilters,
  ...itemInfosFilters,
  ...knownValuesFilters,
  ...d1Filters,
  ...loadoutFilters,
  ...simpleRangeFilters,
  ...overloadedRangeFilters,
  ...simpleFilters,
  ...socketFilters,
  ...statFilters,
  ...locationFilters,
  ...advancedFilters,
];

export const searchConfigSelector = createSelector(
  destinyVersionSelector,
  suggestionsContextSelector,
  buildSearchConfig
);

//
// SearchConfig
//

export interface SearchConfig {
  allFilters: FilterDefinition[];
  /* `is:keyword` filters */
  isFilters: Record<string, FilterDefinition>;
  /* `keyword:value` filters */
  kvFilters: Record<string, FilterDefinition>;
  suggestions: string[];
}

/** Builds an object that describes the available search keywords and filter definitions. */
export function buildSearchConfig(
  destinyVersion: DestinyVersion,
  suggestionsContext: SuggestionsContext = {}
): SearchConfig {
  const suggestions = new Set<string>();
  const isFilters: Record<string, FilterDefinition> = {};
  const kvFilters: Record<string, FilterDefinition> = {};
  const allApplicableFilters: FilterDefinition[] = [];
  for (const filter of allFilters) {
    if (!filter.destinyVersion || filter.destinyVersion === destinyVersion) {
      for (const suggestion of generateSuggestionsForFilter(filter)) {
        suggestions.add(suggestion);
      }
      for (const suggestion of filter.suggestionsGenerator?.(suggestionsContext) ?? []) {
        suggestions.add(suggestion);
      }
      allApplicableFilters.push(filter);
      const filterKeywords = Array.isArray(filter.keywords) ? filter.keywords : [filter.keywords];
      const filterFormats = canonicalFilterFormats(filter.format);
      const hasSimple = filterFormats.some((f) => f === 'simple');
      const hasKv = filterFormats.some((f) => f !== 'simple');

      for (const keyword of filterKeywords) {
        if (hasSimple) {
          if ($DIM_FLAVOR === 'test' && isFilters[keyword]) {
            throw new Error(
              `Conflicting is:${keyword} filter -- only the last inserted filter will work.`
            );
          }
          isFilters[keyword] = filter;
        }
        if (hasKv) {
          if ($DIM_FLAVOR === 'test' && kvFilters[keyword]) {
            throw new Error(
              `Conflicting ${keyword}:value filter -- only the last inserted filter will work.`
            );
          }
          kvFilters[keyword] = filter;
        }
      }
    }
  }

  return {
    allFilters: allApplicableFilters,
    suggestions: Array.from(suggestions),
    isFilters,
    kvFilters,
  };
}
