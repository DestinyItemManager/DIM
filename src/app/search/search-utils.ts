import { FilterDefinition } from './filter-types';
import type { QueryAST } from './query-parser';
import { SearchConfig } from './search-config';

/**
 * loops through collections of strings (filter segments), generating combinations
 *
 * for example, with
 * `[ [a], [b,c], [d,e] ]`
 * as an input, this generates
 *
 * `[ a:, a:b:, a:c:, a:b:d, a:b:e, a:c:d, a:c:e ]`
 */
function expandStringCombinations(stringGroups: string[][], minDepth = 0) {
  const results: string[][] = [];
  for (let i = 0; i < stringGroups.length; i++) {
    const stringGroup = stringGroups[i];
    const stems = results.length ? results[results.length - 1] : undefined;
    const newResults = stringGroup.flatMap((suffix) =>
      stems
        ? stems.map(
            (stem) =>
              (stem ? `${stem}${suffix}` : suffix) + (i === stringGroups.length - 1 ? '' : ':')
          )
        : [`${suffix}:`]
    );
    results.push(newResults);
  }
  return results.slice(minDepth).flat();
}

const operators = ['<', '>', '<=', '>=']; // TODO: add "none"? remove >=, <=?

/**
 * Generates all the possible suggested keywords for the given filter
 *
 * Accepts partial filters with as little as just a "keywords" property,
 * if you want to generate some keywords without a full valid filter
 */
export function generateSuggestionsForFilter(
  filterDefinition: Pick<FilterDefinition, 'keywords' | 'suggestions' | 'format' | 'deprecated'>
) {
  if (filterDefinition.deprecated) {
    return [];
  }

  const { suggestions, keywords } = filterDefinition;
  const thisFilterKeywords = Array.isArray(keywords) ? keywords : [keywords];

  const nestedSuggestions = suggestions === undefined ? [] : [suggestions];

  switch (filterDefinition.format) {
    case 'query':
      return expandStringCombinations([thisFilterKeywords, ...nestedSuggestions]);
    case 'freeform':
      return expandStringCombinations([thisFilterKeywords, []]);
    case 'range':
      return expandStringCombinations([thisFilterKeywords, ...nestedSuggestions, operators]);
    case 'rangeoverload':
      return [
        ...expandStringCombinations([thisFilterKeywords, operators]),
        ...expandStringCombinations([thisFilterKeywords, ...nestedSuggestions]),
      ];
    case 'custom':
      return [];
    default:
      // Pass minDepth 1 to not generate "is:" and "not:" suggestions
      return expandStringCombinations([['is', 'not'], thisFilterKeywords], 1);
  }
}

/**
 * Return whether the query is completely valid - syntactically, and where every term matches a known filter.
 */
export function validateQuery(query: QueryAST, searchConfig: SearchConfig) {
  if (query.error) {
    return false;
  }
  switch (query.op) {
    case 'filter': {
      let filterName = query.type;
      const filterValue = query.args;

      // "is:" filters are slightly special cased
      if (filterName === 'is') {
        filterName = filterValue;
      }

      const filterDef = searchConfig.filters[filterName];
      if (filterDef) {
        // TODO: validate that filterValue is correct
        return true;
      } else {
        return false;
      }
    }
    case 'not':
      return validateQuery(query.operand, searchConfig);
    case 'and':
    case 'or': {
      return query.operands.every((q) => validateQuery(q, searchConfig));
    }
    case 'noop':
      return true;
  }
}
