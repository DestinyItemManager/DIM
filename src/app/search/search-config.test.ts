import { canonicalFilterFormats } from './filter-types';
import { buildFiltersMap } from './search-config';
import { parseAndValidateQuery } from './search-utils';

describe('buildSearchConfig', () => {
  const searchConfig = buildFiltersMap(2);

  test('generates a reasonable filter map', () => {
    expect(Object.keys(searchConfig.isFilters).sort()).toMatchSnapshot('is filters');
    expect(Object.keys(searchConfig.kvFilters).sort()).toMatchSnapshot('key-value filters');
  });

  test('filter formats specify unambiguous formats ', () => {
    /*
     * We have a bunch of filter formats for which `keyword:value`
     * with purely alphabetic values can be valid syntax. Filters should
     * avoid specifying more than one of these.
     * query and freeform filters are sort of the same thing,
     * except queries are exhaustive and freeform aren't. Overloaded
     * range filters can also accept single words as filter value,
     * because `season:worthy` is actually `season:10` and we don't
     * want these to be mistaken for queries or freeforms.
     */

    for (const filter of searchConfig.allFilters) {
      let formats = canonicalFilterFormats(filter.format);

      if (formats.length < 1) {
        throw new Error(`filter ${filter.keywords} has no formats`);
      }

      formats = formats.filter(
        (f) => f === 'query' || f === 'freeform' || (f === 'range' && filter.overload)
      );
      if (formats.length > 1) {
        throw new Error(`filter ${filter.keywords} specifies ambiguous formats ${formats}`);
      }
    }
  });
});

/**
 * The purpose of this test is to verify that `validateQuery` understands the syntax
 * formats in filter definitions, accepts valid queries and rejects invalid syntax.
 * We use well-known filters from the D2 search config for this, but testing all filters
 * exhaustively is not a goal of this test.
 */
describe('validateQuery', () => {
  const searchConfig = buildFiltersMap(2);

  const simpleCases: [filterString: string, valid: boolean][] = [
    ['is:crafted', true],
    ['not:crafted', true],
    ['crafted:is', false],
  ];

  test.each(simpleCases)('is: filter %s - validity %s', (filterString, valid) =>
    expect(parseAndValidateQuery(filterString, searchConfig).valid).toBe(valid)
  );

  const queryCases: [filterString: string, valid: boolean][] = [
    ['tag:favorite', true],
    ['tag:none', true],
    ['tag:any', false],
    ['is:tag', false],
    ['tag:<5', false],
    ['tag:recovery:17', false],
  ];

  test.each(queryCases)('query filter %s - validity %s', (filterString, valid) =>
    expect(parseAndValidateQuery(filterString, searchConfig).valid).toBe(valid)
  );

  const freeformCases: [filterString: string, valid: boolean][] = [
    ['notes:#hashtag', true],
    ['notes:verbatim:colon', true],
    ['notes"with spaces"', true],
    ['notes:<5', true],
    ['is:notes', false],
  ];

  test.each(freeformCases)('freeform filter %s - validity %s', (filterString, valid) =>
    expect(parseAndValidateQuery(filterString, searchConfig).valid).toBe(valid)
  );

  // `masterwork` is a complicated filter with three different formats
  const mixedCases: [filterString: string, valid: boolean][] = [
    ['is:masterwork', true],
    ['masterwork:range', true],
    ['masterwork:<5', true],
    ['masterwork:=5', true],
    ['masterwork:5', true],
    ['masterwork:rnage', false],
    ['masterwork:<range', false],
    ['masterwork:range:5', false],
  ];

  test.each(mixedCases)('mixed filter %s - validity %s', (filterString, valid) =>
    expect(parseAndValidateQuery(filterString, searchConfig).valid).toBe(valid)
  );

  const statCases: [filterString: string, valid: boolean][] = [
    ['stat:recovery:5', true],
    ['stat:recovery:<=5', true],
    ['stat:recovery+discipline:>=5', true],
    ['stat:highest&secondhighest:>20', true],
    ['stat:highest&recovery+strength&strength&range:>20', true],
    ['basestat:range:>50', true],

    ['stat:badstat:>20', false],
    ['stat:badstat&badcombo:>20', false],
    ['stat:too&+many:>20', false],
    ['is:stat', false],
    ['stat:recovery', false],
    ['stat:=5', false],
  ];

  test.each(statCases)('stat filter %s - validity %s', (filterString, valid) =>
    expect(parseAndValidateQuery(filterString, searchConfig).valid).toBe(valid)
  );

  const rangeCases: [filterString: string, valid: boolean][] = [
    ['count:2', true],
    ['count:=2', true],
    ['count:<=2', true],
    ['count:<=2.5', true],

    ['is:count', false],
    ['count:count', false],
    ['count:recovery', false],
    ['count:<=2:=2', false],
    ['count:<2>', false],
  ];

  test.each(rangeCases)('search string %s - validity %s', (filterString, valid) =>
    expect(parseAndValidateQuery(filterString, searchConfig).valid).toBe(valid)
  );

  const overloadRangeCases: [filterString: string, valid: boolean][] = [
    ['season:worthy', true],
    ['season:<=worthy', true],
    ['season:=worthy', true],
    ['season:>=arrival', true],
    ['season:222', true],
    ['season:=10', true],
    ['season:>2.5', true],

    ['is:season', false],
    // DIM used to parse this as `season:11` because `redwar` is `1`...
    ['season:1redwar', false],
    ['season:season', false],
    ['season:arrivals', false],
    ['season:arrivalworthy', false],
  ];

  test.each(overloadRangeCases)('search string %s - validity %s', (filterString, valid) =>
    expect(parseAndValidateQuery(filterString, searchConfig).valid).toBe(valid)
  );
});
