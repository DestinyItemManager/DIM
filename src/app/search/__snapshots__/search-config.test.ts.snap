// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildSearchConfig generates a reasonable filter map 1`] = `
Array [
  "dupe",
  "seasonaldupe",
  "dupelower",
  "count",
  "notes",
  "name",
  "description",
  "perk",
  "perkname",
  "keyword",
  "tagged",
  "tag",
  "hasnotes",
  "common",
  "uncommon",
  "rare",
  "legendary",
  "exotic",
  "white",
  "green",
  "blue",
  "purple",
  "yellow",
  "special",
  "primary",
  "heavy",
  "titan",
  "hunter",
  "warlock",
  "cosmetic",
  "light",
  "haslight",
  "haspower",
  "breaker",
  "kinetic",
  "arc",
  "solar",
  "void",
  "stasis",
  "engrams",
  "lostitems",
  "messages",
  "specialorders",
  "energy",
  "power",
  "helmet",
  "gauntlets",
  "chest",
  "leg",
  "classitem",
  "class",
  "ghost",
  "emblems",
  "ships",
  "vehicle",
  "finishers",
  "seasonalartifacts",
  "clanbanners",
  "consumables",
  "modifications",
  "shaders",
  "autorifle",
  "handcannon",
  "pulserifle",
  "scoutrifle",
  "fusionrifle",
  "sniperrifle",
  "shotgun",
  "machinegun",
  "rocketlauncher",
  "sidearm",
  "sword",
  "grenadelauncher",
  "tracerifle",
  "linearfusionrifle",
  "submachine",
  "smg",
  "bow",
  "transmat",
  "weaponmod",
  "armormod",
  "reptoken",
  "powerfulreward",
  "source",
  "inloadout",
  "stack",
  "year",
  "level",
  "powerlimit",
  "kills",
  "masterwork",
  "energycapacity",
  "season",
  "sunsetsafter",
  "sunsetsin",
  "hascapacity",
  "armor2.0",
  "weapon",
  "armor",
  "equipment",
  "equippable",
  "postmaster",
  "inpostmaster",
  "equipped",
  "transferable",
  "movable",
  "stackable",
  "infusable",
  "infuse",
  "locked",
  "unlocked",
  "new",
  "sunset",
  "randomroll",
  "curated",
  "shaded",
  "hasshader",
  "ornamented",
  "hasornament",
  "hasmod",
  "modded",
  "modslot",
  "holdsmod",
  "stat",
  "basestat",
  "maxstatloadout",
  "maxstatvalue",
  "maxbasestatvalue",
  "maxpower",
  "inleftchar",
  "inmiddlechar",
  "inrightchar",
  "onwrongclass",
  "invault",
  "incurrentchar",
  "id",
  "hash",
]
`;
